include/ginkgo/core/base/index_set.hpp

```diff
@@ -0,0 +1,317 @@
+     * returns the local index in the index set of the provided index set. For
+     * example, `idx_set.get_local_index(0) == 0` `idx_set.get_local_index(4)
+     * == 3` and `idx_set.get_local_index(6) == 4`.
+     *
+     * @note This function returns a scalar value and needs a scalar value.
+     *       It is probably more efficient to use the Array functions that
+     *       take and return arrays which allow for more throughput.
+     *
+     * @param  the global index.
+     * @return  the local index of the element in the index set.
+     */
+    index_type get_local_index(const index_type &global_index) const;
+
+    /**
+     * This is an array version of the scalar function above.
+     *
+     * @param  the local index array.
+     * @return  the global index array from the index set.
+     */
+    Array<index_type> get_global_indices(
+        const Array<index_type> &local_indices) const;
+
+    /**
+     * This is an array version of the scalar function above.
+     *
+     * @param  the global index array.
+     * @return  the local index array from the index set.
+     */
+    Array<index_type> get_local_indices(
+        const Array<index_type> &global_indices) const;
+
+    /**
+     * Checks if the element exists in the index set.
+     *
+     * @return  whether the element exists in the index set.
+     */
+    bool is_element(const index_type &index) const;
+
+    /**
+     * Returns the number of subsets stored in the index set.
+     *
+     * @return  the number of stored subsets.
+     */
+    index_type get_num_subsets() const
+    {
+        return this->subsets_begin_.get_num_elems();
+    }
+
+    /**
+     * Returns a pointer to the beginning indices of the subsets.
+     *
+     * @return  a pointer to the beginning indices of the subsets.
+     */
+    const index_type *get_subsets_begin() const
+    {
+        return this->subsets_begin_.get_const_data();
+    }
+
+    /**
+     * Returns a pointer to the end indices of the subsets.
+     *
+     * @return  a pointer to the end indices of the subsets.
+     */
+    const index_type *get_subsets_end() const
+    {
+        return this->subsets_end_.get_const_data();
+    }
+
+    /**
+     * Returns a pointer to the cumulative indices of the superset of
+     * the subsets.
+     *
+     * @return  a pointer to the cumulative indices of the superset of the
+     *          subsets.
+     */
+    const index_type *get_superset_indices() const
+    {
+        return this->superset_cumulative_indices_.get_const_data();
+    }
+
+private:
+    void populate_subsets(const gko::Array<index_type> &indices);
+
+    std::shared_ptr<const gko::Executor> exec_;
+
+    index_type index_set_id_;
+    index_type index_space_size_;
+    index_type num_stored_indices_;
+    gko::Array<index_type> subsets_begin_;
+    gko::Array<index_type> subsets_end_;
+    gko::Array<index_type> superset_cumulative_indices_;
```

Comment:
For the sake of clarity, could you explain these maybe in the class description, maybe reusing the example you show in `get_local_index`.